File: tests/cases/dns_svc/module.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <iostream>

#include "skullcpp/api.h"
#include "skull_metrics.h"
#include "skull_protos.h"
#include "config.h"

static
void module_init(const skull_config_t* config)
{
    printf("module(test): init\n");
    SKULL_LOG_TRACE("skull trace log test %d", 1);
    SKULL_LOG_DEBUG("skull debug log test %d", 2);
    SKULL_LOG_INFO("1", "skull info log test %d", 3);
    SKULL_LOG_WARN("1", "skull warn log test %d", 4, "ignore, this is test");
    SKULL_LOG_ERROR("1", "skull error log test %d", 5, "ignore, this is test");
    SKULL_LOG_FATAL("1", "skull fatal log test %d", 6, "ignore, this is test");

    skullcpp::Config::instance().load(config);
}

static
void module_release()
{
    printf("module(test): released\n");
    SKULL_LOG_INFO("5", "module(test): released");
}

static
size_t module_unpack(skullcpp::Txn& txn, const void* data, size_t data_sz)
{
    skull_metrics_module.request.inc(1);
    std::cout << "module_unpack(test): data sz: " << data_sz << std::endl;
    SKULL_LOG_INFO("2", "module_unpack(test): data sz:%zu", data_sz);

    // deserialize data to transcation data
    skull::workflow::example& example = (skull::workflow::example&)txn.data();
    example.set_data(data, data_sz);
    return data_sz;
}

static
int _dns_query_cb(skullcpp::Txn& txn, skullcpp::Txn::IOStatus status,
                  const std::string& apiName,
                  const google::protobuf::Message& request,
                  const google::protobuf::Message& response)
{
    std::cout << "dns query result: " << std::endl;
    const skull::service::dns::query_resp& query_resp =
        (const skull::service::dns::query_resp&)response;

    if (!query_resp.code()) {
        // correct
        std::cout << "return ip: " << query_resp.ip() << std::endl;
        SKULLCPP_LOG_INFO("dns_query_cb-1", "query reponse: ip = "
                          << query_resp.ip());
    } else {
        // error ocurred
        std::cout << "dns error reason: " << query_resp.error() << std::endl;
        SKULLCPP_LOG_ERROR("dns_query_cb-2", "query error occurred, reasone: "
            << query_resp.error(), "Network issue?");
    }

    return 0;
}

static
int module_run(skullcpp::Txn& txn)
{
    skull::workflow::example& example = (skull::workflow::example&)txn.data();

    std::cout << "receive data: " << example.data() << std::endl;
    SKULL_LOG_INFO("3", "receive data: %s", example.data().c_str());

    // Call dns service
    skull::service::dns::query_req query_req;
    query_req.set_domain("www.google.com");

    skullcpp::Txn::IOStatus ret =
        txn.serviceCall("dns", "query", query_req, 0, _dns_query_cb);

    std::cout << "ServiceCall ret: " << ret << std::endl;
    return 0;
}

static
void module_pack(skullcpp::Txn& txn, skullcpp::TxnData& txndata)
{
    skull::workflow::example& example = (skull::workflow::example&)txn.data();

    skull_metrics_module.response.inc(1);
    std::cout << "module_pack(test): data sz: " << example.data().length() << std::endl;
    SKULL_LOG_INFO("4", "module_pack(test): data sz:%zu", example.data().length());
    txndata.append(example.data().c_str(), example.data().length());
}

/******************************** Register Module *****************************/
static skullcpp::ModuleEntry module_entry = {
    module_init,
    module_release,
    module_run,
    module_unpack,
    module_pack
};

SKULLCPP_MODULE_REGISTER(&module_entry)

File: tests/cases/multi_modules/module.py
import yaml
import pprint

from skullpy     import *
from skullpy.txn import *

from skull.common import *
from skull.common.proto import *

def module_init(config):
    print("py module init")
    print("config: ", end=' ')
    pprint.pprint(config)

    logger.trace('py module init: trace test')
    logger.debug('py module init: debug test')
    logger.info('1', 'py module init: info test')
    logger.warn('2', 'py module init: warn test', 'no suggestion')
    logger.error('3', 'py module init: error test', 'no solution')
    logger.fatal('4', 'py module init: fatal test', 'no solution')
    return True

def module_release():
    print("py module release")
    return

def module_unpack(txn, data):
    print("data: %s" % data, end=' ')
    logger.info('5', 'receive data: {}'.format(data))
    example_msg = txn.data()
    example_msg.data = data

    return len(data)

def module_pack(txn, txndata):
    mod_metrics = metrics.module()
    mod_metrics.response.inc(1)

    mod_dymetrics = metrics.transaction('test')
    mod_dymetrics.response.inc(1)

    if txn.status() != Txn.TXN_OK:
        txndata.append('error')
        logger.error('6', 'module_pack error', 'no solution')
    else:
        example_msg = txn.data()
        print("pack data: %s" % example_msg.data)
        logger.info('7', 'module_pack: data sz: {}'.format(len(example_msg.data)))

        txndata.append(example_msg.data)

def module_run(txn):
    print("py module_run")

    client = txn.client()
    logger.info('Py PeerInfo', 'peer name: {}, peer port: {}, peer type: {}, peer type name: {}'.format(
        client.name(), client.port(), client.type(), client.typeName()))

    mod_metrics = metrics.module()
    mod_metrics.request.inc(1)

    mod_dymetrics = metrics.transaction('test')
    mod_dymetrics.request.inc(1)

    # Assemble api request
    get_req_msg = service_s1_get_req_pto.get_req()
    get_req_msg.name = "hello"

    # invoke iocall to s1 service
    ret = txn.iocall('s1', 'get', get_req_msg, 0, _api_cb)

    print("iocall ret: {}".format(ret))
    return True

def _api_cb(txn, iostatus, api_name, request_msg, response_msg):
    print("api_cb: iostatus: {}, api_name: {}, request_msg: {}, response_msg: {}".format(
            iostatus, api_name, request_msg, response_msg))

    example_msg = txn.data()
    example_msg.data += (', ' + str(response_msg.response)).encode('UTF-8')
    return True

